### PR TITLE
Branch categorise tags

### PR DESCRIPTION
**Pull request by wang-xinrong**
#67

### CODE REVIEW

File: src/main/java/seedu/address/model/tag/Tag.java [hunk 1]
```diff
@@ -32,12 +32,85 @@ public Tag(String tagName, TagStatus tagStatus) {
...
+        requireNonNull(tagStatus);
+        checkArgument(isValidTagName(tagName), MESSAGE_CONSTRAINTS);
+        TagType tagType = getTagType(tagStatus);
+
+        switch (tagType) {
+        case TUTORIAL:
+            return new TutorialTag(tagName, tagStatus);
+        case ATTENDANCE:
+            return new AttendanceTag(tagName, tagStatus);
+        case ASSIGNMENT:
```
Comment (kaitinghh) [inline src/main/java/seedu/address/model/tag/Tag.java:59]: Should you move return new AssignmentTag(tagName, tagStatus); above to case ASSIGNMENT, and let default return null instead?

File: src/main/java/seedu/address/model/tag/TagStatus.java [hunk 2]
```diff
@@ -7,13 +7,48 @@ public enum TagStatus {
...
     COMPLETE_BAD, // complete after deadline
     INCOMPLETE_GOOD, // incomplete before deadline
-    INCOMPLETE_BAD; // incomplete after deadline
+    INCOMPLETE_BAD, // incomplete after deadline
+
+    PRESENT,
+    ABSENT,
+    ABSENT_WITH_REASON,
+
+    W08, W09, W10, W11, W12, W13,
```
Comment (kaitinghh) [inline src/main/java/seedu/address/model/tag/TagStatus.java:16]: I'm a bit confused with the representation here. Why would the code for tutorial sessions be a valid enum for tagStatus?

File: src/main/java/seedu/address/model/tag/TutorialTag.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package seedu.address.model.tag;
+
+/**
+ * Represents a tutorial tag.
+ */
+public class TutorialTag extends Tag {
```
Comment (kaitinghh) [inline src/main/java/seedu/address/model/tag/TutorialTag.java:6]: What is the difference between TutorialTag and AttendanceTag? I thought we were marking attendance for tutorials?

**Review by kaitinghh — COMMENTED** [review]
Some questions to clarify regarding implementation

File: src/main/java/seedu/address/model/tag/Tag.java (hunk 1 above)
Comment (wang-xinrong) [reply src/main/java/seedu/address/model/tag/Tag.java:59]: sure sure that would work better!

File: src/main/java/seedu/address/model/tag/TagStatus.java (hunk 2 above)
Comment (wang-xinrong) [reply src/main/java/seedu/address/model/tag/TagStatus.java:16]: that was the original plan from our dicussion - using enum to limit the valid tutorial slot value.

however in today's meeting, I would propose to have two enum states for TutorialTag instead: {ASSIGNED, AVAILABLE}, with the first to indicate assigned tutorial for both student and tutor, and the latter for the tutor to indicate other slots they are available. Under this implementation scheme, the tutorial slots would be represented by TagName, and the validity of such TagNames (the tutorial slot has to exist to be inserted) can be checked by an additional construct, maybe a set of valid tutorial codes

File: src/main/java/seedu/address/model/tag/TutorialTag.java (hunk 3 above)
Comment (wang-xinrong) [reply src/main/java/seedu/address/model/tag/TutorialTag.java:6]: Attendance is meant for individual sessions of one tutorial group, such as W1Tutorial, W2Tutorial. On the other hand TutorialTag represents the tutorial session a student is assigned to/a tutor is available during

File: src/main/java/seedu/address/model/tag/TagStatus.java (hunk 2 above)
Comment (kaitinghh) [reply src/main/java/seedu/address/model/tag/TagStatus.java:16]: Yup I think that would be a better idea too

File: src/main/java/seedu/address/model/tag/TutorialTag.java (hunk 3 above)
Comment (kaitinghh) [reply src/main/java/seedu/address/model/tag/TutorialTag.java:6]: Thanks for the clarification!

**Review by wongkj12 — APPROVED** [review]
Implementation looks good.
To clarify the following changes will be needed? (in a future pr):
- Restriction of tag name depending on tag status 
- Students should not be allowed to be marked with "/t TUE08 /ts av"

**Pull request merged**